scanner/decisions/0005-scannerv4-read-cvss-data.md

```diff
@@ -0,0 +1,36 @@
+# 0005 - Adapt Scanner V4 NVD CVSS data updating pipeline to integrate with NVD CVE Api
+
+- **Author(s):** Yi Li <yli3@redhat.com>
+- **Created:** [2023-10-10]
+
+## Status
+
+Accepted
+
+## Context
+
+The National Vulnerability Database (NVD) has announced the retirement of its JSON feeds by December 15, 2023. This creates the need to change our existing CVSS data updater pipeline for Scanner V4, which currently relies on these JSON feeds. 
+
+The NVD JSON feed offers bundled data on a yearly basis, spanning from 2002 to the present year. A benefit is that only two requests are needed in the workflow for each year (one for the meta file and one for the data bundle). However, a limitation is the absence of filtering capabilities, such as by CVSS score or modification date. In contrast, the NVD API allows data retrieval through specific HTTP URL parameters, such as publish date, modified date, or CVSS score, enabling more precise data acquisition.
+
+## Decision
+
+The updated NVD CVSS GitHub Workflow (CI) has transitioned to sourcing data from the NVD CVE API, rather than the NVD Json feeds, ensuring the timely update of the NVD CVSS data bundle in Google Storage. To fetch data, the workflow downloads based on both the start index and CVSS v3 severity. For instance, a command such as curl "https://services.nvd.nist.gov/rest/json/cves/2.0?cvssV3Severity=CRITICAL&startIndex=0" > critical-1.json would be used. For every severity level, the workflow iteratively exhaust the startIndex to capture all associated data. And there are four severity levels in total. After all json files downloaded, they will be compress to one data bundle as one single zip file.
```

Comment:
nit: `rather than the NVD Json feeds` -> `rather than the NVD JSON feeds`

Reply:
Can you reformat the command to be more markdown friendly? For example:

To fetch data, the workflow downloads based on both the start index and CVSS v3 severity. For instance, a command such as the following:

```
curl "https://services.nvd.nist.gov/rest/json/cves/2.0?cvssV3Severity=CRITICAL&startIndex=0" > critical-1.json
```

Reply:
`For every severity level, the workflow iteratively exhaust the startIndex to capture all associated data.`

What does this mean? Is this just saying we loop through calls to this endpoint via updating `startIndex`?

Also, how do we know what it is done? What indicates this? A status code? Does this content of this endpoint tell us how many vulns exist which fit the criteria?

Also, nit: `exhausts`

Reply:
`And there are four severity levels in total.` What are the severity levels? Can you list them? I think this sentence isn't needed. You can just add the list of severities to the sentence "For every severity level..."

Also nit: For each severity level

Reply:
nit: `After all json files downloaded, they will be compress to one data bundle as one single zip file.` -> `After all JSON files are downloaded, they will be compressed into a single ZIP file.`

Also, how did we decide on ZIP? Have we considered a tar.gz or zstd?

Reply:
I took a look at how ClairCore fetches NVD data, and it assumes there is a file per year. If we use ClairCore's enricher, how will we reformat the data to fit this model? Similarly, ClairCore's enricher assumes the data is in the 1.1 JSON feed format, which is different from the 2.0 API format. I think we will need to use our own, modified version of ClairCore's CVSS enricher. This also gives us the flexibility to get more data out of NVD that ClairCore currently gets, as we may need more than what vanilla ClairCore offers

Reply:
"a file per year" is the json feeds that's going to be retired.
I concur with the idea of developing our own enricher. It's essential to ensure that our enricher can parse data from the Google bucket. I'll incorporate this detail into the ADR.

Reply:
What is stopping us from downloading the CVEs year by year? The API appears to support this.  Or downloading all vulns (paging through the indexes) and creating the yearly files ourselves.

Reply:
I am not sure what's the benefit of the downloading data for every year.
We lost the conveniences of excluding data without cvss v3 and we still have the constraint of pagination. 

The yearly JSON feeds doesn't have pagination limitation so that was something nice to have. 

What do you think? @dcaravel

Reply:
> `For every severity level, the workflow iteratively exhaust the startIndex to capture all associated data.`
> 
> What does this mean? Is this just saying we loop through calls to this endpoint via updating `startIndex`?
> 
> Also, how do we know what it is done? What indicates this? A status code? Does this content of this endpoint tell us how many vulns exist which fit the criteria?
> 
> Also, nit: `exhausts`
In each json that's been downloaded, there are these fields to determine how many results there are in total, where the index starts and how many results per page.
resultsPerPage":2000,"startIndex":0,"totalResults":21611

Reply:
> nit: `After all json files downloaded, they will be compress to one data bundle as one single zip file.` -> `After all JSON files are downloaded, they will be compressed into a single ZIP file.`
> 
> Also, how did we decide on ZIP? Have we considered a tar.gz or zstd?


Zip might not be most ideal, I think we can either use `tar -xf all_json_files.tar.gz`
 or `tar -xf all_json_files.tar.xz`